sycl/include/sycl/ext/oneapi/experimental/cuda/non_uniform_algorithms.hpp

```diff
@@ -101,86 +102,12 @@ masked_reduction_cuda_shfls(Group g, T x, BinaryOperation binary_op,
                             const uint32_t MemberMask) {
   for (int i = g.get_local_range()[0] / 2; i > 0; i /= 2) {
     T tmp;
-    if constexpr (std::is_same_v<T, double>) {
-      int x_a, x_b;
-      asm volatile("mov.b64 {%0,%1},%2;" : "=r"(x_a), "=r"(x_b) : "d"(x));
-      auto tmp_a = __nvvm_shfl_sync_bfly_i32(MemberMask, x_a, -1, i);
-      auto tmp_b = __nvvm_shfl_sync_bfly_i32(MemberMask, x_b, -1, i);
-      asm volatile("mov.b64 %0,{%1,%2};" : "=d"(tmp) : "r"(tmp_a), "r"(tmp_b));
-    } else if constexpr (std::is_same_v<T, long> ||
-                         std::is_same_v<T, unsigned long>) {
-      int x_a, x_b;
-      asm volatile("mov.b64 {%0,%1},%2;" : "=r"(x_a), "=r"(x_b) : "l"(x));
-      auto tmp_a = __nvvm_shfl_sync_bfly_i32(MemberMask, x_a, -1, i);
-      auto tmp_b = __nvvm_shfl_sync_bfly_i32(MemberMask, x_b, -1, i);
-      asm volatile("mov.b64 %0,{%1,%2};" : "=l"(tmp) : "r"(tmp_a), "r"(tmp_b));
-    } else if constexpr (std::is_same_v<T, half>) {
-      short tmp_b16;
-      asm volatile("mov.b16 %0,%1;" : "=h"(tmp_b16) : "h"(x));
-      auto tmp_b32 = __nvvm_shfl_sync_bfly_i32(
-          MemberMask, static_cast<int>(tmp_b16), -1, i);
-      asm volatile("mov.b16 %0,%1;"
-                   : "=h"(tmp)
-                   : "h"(static_cast<short>(tmp_b32)));
-    } else if constexpr (std::is_same_v<T, float>) {
-      auto tmp_b32 =
-          __nvvm_shfl_sync_bfly_i32(MemberMask, __nvvm_bitcast_f2i(x), -1, i);
-      tmp = __nvvm_bitcast_i2f(tmp_b32);
-    } else {
-      tmp = __nvvm_shfl_sync_bfly_i32(MemberMask, x, -1, i);
-    }
+    CUDA_SHFL_SYNC(tmp, MemberMask, x, i, 0x1f, bfly_i32)
```

Comment:
It looks like now the nvvm built-ins will be called with `i` instead of `-1` and `0x1f` instead of `i`, was that the bug you fixed?

Reply:
This change is not a bug fix: both of the two ways of doing xor shuffles for such a reduction are correct. I switched to the more standard way that is described here: https://people.maths.ox.ac.uk/~gilesm/cuda/lecs/lec4.pdf

Reply:
For further reference the full impl of sycl_xor_shfl described there matches this function:
https://github.com/oneapi-src/SYCLomatic/pull/1342/files#diff-1811724d606e4bab73b204eb3834147684334bf4000e2b63c4203c4563d4c4e5R542